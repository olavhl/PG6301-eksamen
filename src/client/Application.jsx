import React from "react";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

export function Application() {
  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>
          <button>Home</button>
        </Link>
      </header>
      <Switch>
        <Route exact path={"/login"}>
          <h1>Login</h1>
        </Route>
        <Route path={"/login/callback"}>
          <h1>Login Callback</h1>
        </Route>
        <Route exact path={"/"}>
          <h1>Welcome to the chatting page</h1>
          <ul>
            <li>
              <Link to={"/login"}>Login</Link>
            </li>
          </ul>
        </Route>
        <Route>
          <h1>Page not found...</h1>
        </Route>
      </Switch>
    </BrowserRouter>
  );
}
