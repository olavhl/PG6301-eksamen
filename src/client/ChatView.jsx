import React, { useState } from "react";

export function ChatView({ onSendMessage, chat }) {
  const [message, setMessage] = useState("");
  // Feedback
  const [feedBack, setFeedBack] = useState("");

  function handleSubmitChat(e) {
    e.preventDefault();

    if (message === "") {
      setFeedBack("Write the message you want to send!");
      return;
    }

    onSendMessage(message);
    setMessage("");
  }

  return (
    <div className={"chat"}>
      <header>
        <h1>Chat</h1>
      </header>
      <main>
        {chat?.map((message, index) => (
          <div key={`${message}__${index}`}>{message}</div>
        ))}
      </main>
      {feedBack && <div className={"feedback"}>{feedBack}</div>}
      <footer>
        <form onSubmit={handleSubmitChat}>
          <input
            type="text"
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Send</button>
        </form>
      </footer>
    </div>
  );
}
