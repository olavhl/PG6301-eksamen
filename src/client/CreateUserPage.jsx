import React, { useState } from "react";
import { InputField } from "./components/InputField";

export function CreateUserPage() {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");

  async function handleSubmit(e) {
    e.preventDefault();
  }

  return (
    <div>
      <h1>Create user</h1>
      <form onSubmit={handleSubmit}>
        <InputField
          value={firstName}
          label={"First name"}
          onChangeValue={setFirstName}
        />
        <InputField
          value={lastName}
          label={"Last name"}
          onChangeValue={setLastName}
        />
        <InputField value={email} label={"Email"} onChangeValue={setEmail} />
        <button>Submit</button>
      </form>
    </div>
  );
}
