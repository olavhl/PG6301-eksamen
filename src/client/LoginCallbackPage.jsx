import React, { useEffect } from "react";
import { useHistory } from "react-router";

export function LoginCallbackPage({ onAccessToken }) {
  const hash = Object.fromEntries(
    new URLSearchParams(window.location.hash.substring(1))
  );

  const history = useHistory();

  useEffect(() => {
    const { access_token } = hash;
    onAccessToken(access_token);
    history.push("/");
  }, []);

  return <h1>Login Callback</h1>;
}
