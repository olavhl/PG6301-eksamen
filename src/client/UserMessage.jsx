import React, { useState } from "react";
import { useLoading } from "./lib/useLoading";
import { ErrorView } from "./components/ErrorView";
import { LoadingView } from "./components/LoadingView";
import { UserCreateUserButton } from "./components/UserCreateUserButton";
import { useParams } from "react-router";

export function UserMessage({ userApi }) {
  const [messages, setMessages] = useState("");
  // Feedback
  const [feedBack, setFeedBack] = useState("");

  const { id } = useParams();
  const { data: user, error, loading, reload } = useLoading(
    async () => await userApi.getUser(id),
    [id]
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading) {
    return <LoadingView />;
  }

  // Checking if there are any users in the UserArray in server
  if (user.length === 0) {
    return <UserCreateUserButton />;
  }

  async function handleSendMessage(e) {
    e.preventDefault();

    if (messages === "") {
      setFeedBack(`Write a message to ${user.firstName}`);
      return;
    }

    await userApi.sendMessage(id, messages);
    // Refreshing site to be able to see all messages
    window.location.reload(false);
  }

  return (
    <div>
      {user && (
        <h1>
          Messages: {user.firstName} {user.lastName}
        </h1>
      )}

      {user.messages &&
        user.messages.map((message, index) => (
          <p key={`${message}__${index}`}>You: {message}</p>
        ))}

      <form onSubmit={handleSendMessage}>
        <input
          type="text"
          value={messages}
          onChange={(e) => setMessages(e.target.value)}
        />
        <button>Send</button>
      </form>
      <br />
      {feedBack && <div className={"feedback"}>{feedBack}</div>}
    </div>
  );
}
