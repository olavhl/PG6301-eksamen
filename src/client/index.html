<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Messaging System</title>
    <link rel="stylesheet" href="style.css" type="text/css">
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
