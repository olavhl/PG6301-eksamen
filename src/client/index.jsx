import React from 'react';
import ReactDOM from 'react-dom';

function Application() {
    return <h1>Hello</h1>;
}

ReactDOM.render(<Application/>, document.getElementById("app"));