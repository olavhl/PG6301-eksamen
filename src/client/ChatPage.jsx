import React, { useEffect, useState } from "react";
import { ChatView } from "./ChatView";

export function ChatPage() {
  const [chat, setChat] = useState([]);
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://localhost:3000/chat");
    ws.onopen = (event) => {
      console.log("opened", event);
    };

    ws.onmessage = (event) => {
      console.log("message", event);
      setChat((chat) => [...chat, event.data]);
    };

    ws.onclose = (event) => {
      console.log("close", event);
    };

    setWs(ws);
  }, []);

  return <ChatView chat={chat} onSendMessage={(message) => ws.send(message)} />;
}
