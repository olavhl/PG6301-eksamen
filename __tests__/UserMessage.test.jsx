import React from "react";
import ReactDOM from "react-dom";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter, Route } from "react-router";
import { UserMessage } from "../src/client/UserMessage";

async function renderToDom(component, url = "/user/99/message") {
  const container = document.createElement("div");
  await act(async () => {
    await ReactDOM.render(
      <MemoryRouter initialEntries={[url]}>
        <Route path={"/user/:id/message"}>{component}</Route>
      </MemoryRouter>,
      container
    );
  });

  return container;
}

describe("User message page", () => {
  it("loads user", async () => {
    const getUser = () => ({
      firstName: "Ola",
      lastName: "Nordmann",
      email: "ola@nordmann.no",
    });
    const container = await renderToDom(<UserMessage userApi={{ getUser }} />);

    expect(container.querySelector("h1").textContent).toEqual(
      "Messages: Ola Nordmann"
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  it("show user message error", async () => {
    const getUser = async () => {
      throw Error("Test is failing");
    };
    const container = await renderToDom(<UserMessage userApi={{ getUser }} />);

    expect(container.querySelector("div").textContent).toEqual(
      "Something went wrongError: Test is failingTry again"
    );
  });

  it("show user messages feedback", async () => {
    const getUser = () => ({
      firstName: "Ola",
      lastName: "Nordmann",
      email: "ola@nordmann.no",
    });
    const container = await renderToDom(<UserMessage userApi={{ getUser }} />);

    Simulate.submit(container.querySelector("form"));
    expect(container.querySelector(".feedback").textContent).toEqual(
      "Write a message to Ola"
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  it("show no-user-btn", async () => {
    const getUser = () => [];
    const container = await renderToDom(<UserMessage userApi={{ getUser }} />);

    expect(container.querySelector("h3").textContent).toEqual(
      "You dont have any users/friends :("
    );
    expect(container.innerHTML).toMatchSnapshot();
  });
});
