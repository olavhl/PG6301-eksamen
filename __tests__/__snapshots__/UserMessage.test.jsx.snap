// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User message page loads user 1`] = `"<div><h1>Messages: Ola Nordmann</h1><form><input type=\\"text\\" value=\\"\\"><button>Send</button></form><br></div>"`;

exports[`User message page show no-user-btn 1`] = `"<div><h3>You dont have any users/friends :(</h3><a href=\\"/user/create\\"><button>Create user</button></a></div>"`;

exports[`User message page show user messages feedback 1`] = `"<div><h1>Messages: Ola Nordmann</h1><form><input type=\\"text\\" value=\\"\\"><button>Send</button></form><br><div class=\\"feedback\\">Write a message to Ola</div></div>"`;
